presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,421 @@
+                subplan.accept(this, context); // visit child
+            }
+
+            return null;
+        }
+
+        @Override
+        public Void visitTableWriter(TableWriterNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+            source.accept(this, context); // visit child
+
+            return null;
+        }
+
+        @Override
+        public Void visitDelete(DeleteNode node, Void context)
+        {
+            PlanNode source = node.getSource();
+            source.accept(this, context); // visit child
+
+            return null;
+        }
+
+        @Override
+        public Void visitMetadataDelete(MetadataDeleteNode node, Void context)
+        {
+            return null;
+        }
+
+        @Override
+        public Void visitTableFinish(TableFinishNode node, Void context)
+        {
+            node.getSource().accept(this, context); // visit child
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            for (int i = 0; i < node.getSources().size(); i++) {
+                PlanNode subplan = node.getSources().get(i);
+                subplan.accept(this, context); // visit child
+            }
+            checkMappingTypeMatching(node.getSymbolMapping());
+
+            return null;
+        }
+
+        @Override
+        public Void visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            node.getSource().accept(this, context); // visit child
+
+            return null;
+        }
+
+        private void checkMappingTypeMatching(ListMultimap<Symbol, Symbol> symbolMapping)
+        {
+            for (Symbol symbol : symbolMapping.keySet()) {
+                List<Symbol> symbolList = symbolMapping.get(symbol);
+                Type type = types.get(symbol);
+                for (Symbol element : symbolList) {
+                    verifyTypeSignature(symbol, type.getTypeSignature(), types.get(element).getTypeSignature());
+                }
+            }
+        }
+
+        private void checkSignatureTypeMatching(Map<Symbol, Signature> functions)
+        {
+            for (Map.Entry<Symbol, Signature> entry : functions.entrySet()) {
+                TypeSignature typeSignature = types.get(entry.getKey()).getTypeSignature();
+                TypeSignature functionTypeSignature = entry.getValue().getReturnType();
+                verifyTypeSignature(entry.getKey(), typeSignature, functionTypeSignature);
+            }
+        }
+
+        private void checkFunctionCallTypeMatching(Map<Symbol, FunctionCall> functionCalls)
+        {
+            for (Map.Entry<Symbol, FunctionCall> entry : functionCalls.entrySet()) {
+                Type type = types.get(entry.getKey());
+                Type expressionType = getExpressionTypes(session, metadata, sqlParser, types, entry.getValue()).get(entry.getValue());
```

Comment:
Thank @kokosing for pointing it out.

Actually I have also considered getting types information from `Analysis` since `getExpressionTypes` is quite expensive. But this checker is run after all the optimizers that might change or add some internal expressions whose type information could not be retrieved from `Analysis`.

Another approach is to first try retrieve it from `Analysis`, if there is not, then we go to `getExpressionTypes`, but it requires passing `Analysis`, of course it might not be a big deal.

@kokosing which approach do you think is better?